substrate/client/rpc/src/lib.rs

```diff
@@ -45,3 +45,147 @@ pub mod testing;
 
 /// Task executor that is being used by RPC subscriptions.
 pub type SubscriptionTaskExecutor = std::sync::Arc<dyn sp_core::traits::SpawnNamed>;
+
+/// JSON-RPC helpers.
+pub mod utils {
+	use crate::SubscriptionTaskExecutor;
+	use futures::{
+		future::{self, Either, Fuse, FusedFuture},
+		Future, FutureExt, Stream, StreamExt,
+	};
+	use jsonrpsee::{PendingSubscriptionSink, SubscriptionMessage, SubscriptionSink};
+	use sp_runtime::Serialize;
+	use std::collections::VecDeque;
+
+	/// A simple bounded VecDeque.
+	struct BoundedVecDeque<T> {
+		inner: VecDeque<T>,
+		max_cap: usize,
+	}
+
+	impl<T> BoundedVecDeque<T> {
+		/// Create a new bounded VecDeque.
+		fn new(max_cap: usize) -> Self {
+			Self { inner: VecDeque::with_capacity(max_cap), max_cap }
+		}
+
+		fn push_back(&mut self, item: T) -> Result<(), ()> {
+			if self.inner.len() >= self.max_cap {
+				Err(())
+			} else {
+				self.inner.push_back(item);
+				Ok(())
+			}
+		}
+
+		fn pop_front(&mut self) -> Option<T> {
+			self.inner.pop_front()
+		}
+	}
+
+	/// Feed items to the subscription from the underlying stream.
+	///
+	/// It's possible configure how many items from the stream
+	/// are allowed to be kept in memory until the subscription is dropped.
+	///
+	/// This is needed because the underlying streams in substrate are
+	/// unbounded.
+	pub async fn pipe_from_stream<S, T>(
+		pending: PendingSubscriptionSink,
+		mut stream: S,
+		max_cap: usize,
+	) where
+		S: Stream<Item = T> + Unpin + Send + 'static,
+		T: Serialize + Send + 'static,
+	{
+		let mut buf = BoundedVecDeque::new(max_cap);
+		let accept_fut = pending.accept();
+
+		futures::pin_mut!(accept_fut);
+
+		// Poll the stream while waiting for the subscription to be accepted
+		//
+		// If the `max_cap` is exceeded then the subscription is dropped.
+		let sink = loop {
```

Comment:
Ideally, we would only initialize the stream after the subscription was "accepted" but
that would be a breaking change and I don't want to break it here.

Once the subscription has been "accepted" we can't "reject" except sending a special subscription notification for it.

Reply:
That would def be an improvement I feel; something to try after this merges perhaps?